### PR TITLE
feat: Open Telemetry Publish Side Support

### PR DESCRIPTION
**Pull request by mukund-ananthu**
Publish Side Open Telemetry support. Related Issue: https://github.com/googleapis/python-pubsub/issues/543

### CODE REVIEW

**Comment by mukund-ananthu** [comment]
@hongalex PTAL Publish Side Open Telemetry code. Adding @parthea for Python review

**Review by parthea — COMMENTED** [review]
Added initial review

File: google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py [hunk 1]
```diff
@@ -115,15 +117,15 @@ def _create_batch(
...
+        wrapper: PublishMessageWrapper,
         retry: "OptionalRetry" = gapic_v1.method.DEFAULT,
         timeout: "types.OptionalTimeout" = gapic_v1.method.DEFAULT,
     ) -> "futures.Future":
         """Batch message into existing or new batch.
 
         Args:
-            message:
-                The Pub/Sub message.
+            wrapper:
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py:127]: Could you elaborate why you think this is a breaking change. The publish method here is not the publisher client.publish() method used by customers for publishing to the service:
https://github.com/googleapis/python-pubsub/blob/eafc1c87e0881cec5d1b689b24798a86eb286957/google/cloud/pubsub_v1/publisher/client.py#L269

But an internal method used by the library itself .

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    def __eq__(self, other):  # pragma: NO COVER
+        """Used for pytest asserts to compare two PublishMessageWrapper objects with the same message"""
+        if isinstance(self, other.__class__):
+            return self.message == other.message
+        return False
+
+    def start_create_span(self, topic: str, ordering_key: str) -> None:
+        tracer = trace.get_tracer(self._OPEN_TELEMETRY_TRACER_NAME)
+        topic_short_name = topic.split("/")[3]
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py:61]: We could do this as a precautionary measure, but the statement as is matches format and length of this topic specified in the public documentation: https://cloud.google.com/pubsub/docs/publish-receive-messages-client-library#publish_messages

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                },
+            )
+            self._create_span: trace.Span = create_span
+            TraceContextTextMapPropagator().inject(
+                carrier=self._message,
+                setter=OpenTelemetryContextSetter(),
+            )
+
+    def end_create_span(self, exc: Optional[BaseException] = None) -> None:
+        if self._create_span is None:  # pragma: NO COVER
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py:89]: The tests already exercise the expected code path, i.e if we are calling end_create_span on the publish wrapper, the code must have started it previously. If it hasn't been started, then it is an error in the code. The pragma NO Cover is for test coverage for the unexpected case of the create_span being None, when its not supposed to. If the pragma no cover is replaced with a test, the test would have to be mocking this scenario since it would not occur / isn't currently occurring in the code. 

But I have created a bug to just evaluate usage of pragma no cover across the repo here: https://github.com/googleapis/python-pubsub/issues/1249

File: google/cloud/pubsub_v1/open_telemetry/context_propagation.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+class OpenTelemetryContextSetter(Setter):
+    """
+    Used by Open Telemetry for context propagation.
+    """
+
+    def set(self, carrier: gapic_types.PubsubMessage, key: str, value: str) -> None:
+        """
+        Injects trace context into Pub/Sub message attributes with
+        "googclient_" prefix.
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/context_propagation.py:28]: Done

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hongalex) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: In other languages, it is possible that if the user creates a fake of the pubsub service, they might return a topic not in the format that's specified in the public documentation. This might be worth guarding around.

[the remark this answers is not in the document]
Comment (hongalex) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: Duplicate of above. Closing.

File: tests/unit/pubsub_v1/publisher/test_publisher_client.py [hunk 5]
```diff
@@ -129,6 +135,222 @@ def test_init_w_custom_transport(creds):
...
+
+    spans = span_exporter.get_finished_spans()
+    # Span 1 = Publisher Flow Control Span of first publish
+    # Span 2 = Publisher Batching Span of first publish
+    # Span 3 = Publisher Flow Control Span of second publish(raises FlowControlLimitError)
+    # Span 4 = Publish Create Span of second publish(raises FlowControlLimitError)
+    assert len(spans) == 4
+
+    failed_flow_control_span = spans[2]
+    finished_publish_create_span = spans[3]
```
Comment (hongalex) [inline tests/unit/pubsub_v1/publisher/test_publisher_client.py:297]: Is it possible these spans might be created in a different order?

File: google/cloud/pubsub_v1/open_telemetry/context_propagation.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2017, Google LLC All rights reserved.
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/context_propagation.py]: Done

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        with tracer.start_as_current_span(
+            name=f"{topic_short_name} create",
+            attributes={
+                "messaging.system": self._OPEN_TELEMETRY_MESSAGING_SYSTEM,
+                "messaging.destination.name": topic_short_name,
+                "code.function": "publish",
+                "messaging.gcp_pubsub.message.ordering_key": ordering_key,
+                "messaging.operation": "create",
+                "gcp.project_id": topic.split("/")[1],
+                "messaging.message.body.size": sys.getsizeof(self._message.data),
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: Done

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: Adding an assert on the length of the topic. 

I checked what happens if the  users tries to publish messages in a format that's not topic=`projects/projectID/topics/topicID` in the current state of the library(i.e without Open Telemetry). It fails as expected:

```
Published messages with error handler to mukund-topic.
exception calling callback for <Future at 0x7f9b5bb51fd0 state=finished raised InvalidArgument>
Traceback (most recent call last):
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 340, in _invoke_callbacks
    callback(self)
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/publish.py", line 21, in callback
    print(publish_future.result(timeout=60))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/futures.py", line 66, in result
    return super().result(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py", line 274, in _commit
    response = self._client._gapic_publish(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/client.py", line 267, in _gapic_publish
    return super().publish(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/pubsub_v1/services/publisher/client.py", line 1067, in publish
    response = rpc(
               ^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 293, in retry_wrapped_func
    return retry_target(
           ^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 153, in retry_target
    _retry_error_helper(
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_base.py", line 212, in _retry_error_helper
    raise final_exc from source_exc
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 144, in retry_target
    result = target()
             ^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 78, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InvalidArgument: 400 Invalid resource name given (name=mukund-topic). Refer to https://cloud.google.com/pubsub/docs/pubsub-basics#resource_names for more information. [detail: "Could not parse topic \'mukund-topic\'"
]
```

This is because the topic supplied is bubbled down all the way to gapic publish, which on attempting to publish to the backend with a topic specified in any other format, would get the above error.

File: google/cloud/pubsub_v1/open_telemetry/__init__.py
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/__init__.py:1]: Done

File: tests/unit/pubsub_v1/publisher/test_publisher_client.py (hunk 5 above)
Comment (mukund-ananthu) [reply tests/unit/pubsub_v1/publisher/test_publisher_client.py:297]: No, because I create / end them in a particular order.

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            return self.message == other.message
+        return False
+
+    def start_create_span(self, topic: str, ordering_key: str) -> None:
+        tracer = trace.get_tracer(self._OPEN_TELEMETRY_TRACER_NAME)
+        topic_short_name = topic.split("/")[3]
+        with tracer.start_as_current_span(
+            name=f"{topic_short_name} create",
+            attributes={
+                "messaging.system": self._OPEN_TELEMETRY_MESSAGING_SYSTEM,
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: done.

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hongalex) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py:61]: It would be incorrect implementation of the fake, correct. It's something I had dealt with in Go (where some of our tests allows creating a topic without a valid project ID). Anyways I don't feel strongly about this so I'm ok with leaving as is.

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hongalex) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py:65]: Since the python semconv is not stable, and it is not common to pin to specific versions of dependencies, we will not be using the semconv constants.

File: google/cloud/pubsub_v1/publisher/_batch/thread.py [hunk 9]
```diff
@@ -226,6 +239,37 @@ def _start_commit_thread(self) -> None:
...
+    def _start_publish_rpc_span(self) -> None:
+        tracer = trace.get_tracer(self._OPEN_TELEMETRY_TRACER_NAME)
+        links = []
+
+        for wrapper in self._message_wrappers:
+            span = wrapper.create_span
+            # Add links only for sampled spans.
+            if span.get_span_context().trace_flags.sampled:
+                links.append(trace.Link(span.get_span_context()))
+        topic_short_name = self._topic.split("/")[3]
```
Comment (hongalex) [inline google/cloud/pubsub_v1/publisher/_batch/thread.py:251]: nit: use the same assert as above here and for project ID?

**Review by hongalex — APPROVED** [review]
one nit but otherwise LGTM

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py:61]: Mentioning in case this got missed - I have added an assert to guard against this case `Adding an assert on the length of the topic.`.

File: google/cloud/pubsub_v1/publisher/_batch/thread.py (hunk 9 above)
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/publisher/_batch/thread.py:252]: Done.

File: google/cloud/pubsub_v1/publisher/_batch/thread.py [hunk 10]
```diff
@@ -85,6 +91,9 @@ class Batch(base.Batch):
             timeout is used.
     """
 
+    _OPEN_TELEMETRY_TRACER_NAME: str = "google.cloud.pubsub_v1.publisher"
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/publisher/_batch/thread.py]: Marking this resolve, but please feel free to reopen.

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (parthea) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: > I'd just end up mocking the wrapper to create error scenarios where end_create_span is called when there is no span to end.

Please add a TODO to add these tests as part of follow on work

File: google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (parthea) [reply google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py:127]: It's still good to include this in the release notes, even though it was an internal change. We don't need to label is as "breaking change" but we should mention it in the release notes

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: I just ended up adding the tests with the unexpected scenarios themselves (which was what I would have had to do as part of the TODO)

File: google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py:127]: Okay

File: setup.py [hunk 11]
```diff
@@ -45,6 +45,8 @@
     "protobuf>=3.20.2,<6.0.0dev,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5",
     "grpc-google-iam-v1 >= 0.12.4, < 1.0.0dev",
     "grpcio-status >= 1.33.2",
+    "opentelemetry-api",
+    "opentelemetry-sdk",
 ]
 extras = {"libcst": "libcst >= 0.3.10"}
```
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply setup.py]: Adding opentelemetry as an optional dependency would require peppering the code base with 

```
try:
   from opentelemetry import X, Y, Z
   open_telemetry_imported = True
except ImportError:
   open_telemetry_imported = False

if open_telemetry_imported:
   // use open telemetry
```

And this gets messier with subclasses of opentelemetry classes like google/cloud/pubsub_v1/open_telemetry/context_propagation.py

where we have

```

from opentelemetry.propagators.textmap import Setter

class OpenTelemetryContextSetter(Setter):
   // class definition
```

And OpenTelemetryContextSetter is imported into other files in the repository. I would rather prefer just adding opentelemetry as a non-optional dependency. I think the addition of one dependency, which negates the need to pepper  try/except + if import clauses would be an okay tradeoff. Plmk if that works for you @hongalex @parthea

File: google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py]: Resolving this comment since I've addressed it. But please feel free to reopen if anything additional needs to be done here.

File: setup.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mukund-ananthu) [reply setup.py:51]: Discussed this offline. We decided to stick with hard dependencies since the opentelemetry dependency is only 2.8M (excluding transitive dependencies). There isn't too much of a size increase in the client library. The advantages is we avoid convoluted try / except blocks and if import_success / else blocks in the code base.
